### PR TITLE
perf!: use icicle msm

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR integrates [icicle](https://github.com/ingonyama-zk/icicle) MSM for improved performance!

Previously, the MSM GPU interface accepted input directly from GPU memory, which complicated the use of MSM GPU for the KZG commitment scheme. While we could implement this for previous algorithms, such as Bellman MSM and CUZK, they were slow when dealing with non-random scalar distributions. The advantages of using icicle are twofold: (1) it supports input from CPU memory, and (2) it performs well even with non-random scalar distributions. Consequently, this PR removes the previous implementations and adopts icicle exclusively.

### CODE REVIEW

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

File: tachyon/c/math/elliptic_curves/generator/msm_gpu.h.tpl [hunk 1]
```diff
@@ -19,12 +19,12 @@ typedef struct tachyon_%{type}_g1_msm_gpu* tachyon_%{type}_g1_msm_gpu_ptr;
 %{extern_c_front}
 
 /**
- * @brief Creates a new GPU-accelerated MSM context for the G1 group with a specified polynomial degree and algorithm choice.
+ * @brief Creates a new GPU-accelerated MSM context for the G1 group with a specified polynomial degree.
  * @param degree The degree of the polynomial for the MSM operation. This parameter may influence the allocation and optimization strategy.
  * @param algorithm The specific algorithm to use for MSM on the GPU. Different algorithms may be better suited to different scenarios.
```
Comment (ashjeong) [inline tachyon/c/math/elliptic_curves/generator/msm_gpu.h.tpl:24]: Please delete this as well!

File: tachyon/c/math/elliptic_curves/msm/BUILD.bazel [hunk 2]
```diff
@@ -11,12 +11,7 @@ package(default_visibility = ["//visibility:public"])
 
 filegroup(
     name = "msm_hdrs",
```
Comment (ashjeong) [inline tachyon/c/math/elliptic_curves/msm/BUILD.bazel:13]: Can this just be deleted then?

File: tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h [hunk 3]
```diff
@@ -5,13 +5,15 @@
 
 #include "third_party/icicle/include/fields/id.h"
 
+#include "tachyon/device/gpu/gpu_memory.h"
+#include "tachyon/math/elliptic_curves/affine_point.h"
```
Comment (ashjeong) [inline tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h:10]: Is `jacobian_point.h` unneeded?

**Review by ashjeong — CHANGES_REQUESTED** [review]
Could you rebase to remove the first 5 commits?

c300a7020a8a9002fb96a1f156389969b52f1fa2
Please change to “…different types of commitments”

**Comment by TomTaehoonKim** [comment]
How about adding GPU benchmark as well?

**Review by GiDeokKim — COMMENTED** [review]
First of all, I only reviewed the Bazel part related to icicle.

File: tachyon/c/math/elliptic_curves/msm/BUILD.bazel (hunk 2 above)
Comment (chokobole) [reply tachyon/c/math/elliptic_curves/msm/BUILD.bazel:13]: Yes, i'll delete it then :)

File: tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h (hunk 3 above)
Comment (chokobole) [reply tachyon/math/elliptic_curves/msm/algorithms/icicle/icicle_msm.h:10]: This file has been refactored to use only `ProjectivePoint`, `jacobian_point.h` is not needed and `affine_point.h` also should be removed.

File: third_party/icicle/icicle.BUILD [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            "FIELD_ID=BW6_761",
+            "CURVE_ID=BW6_761",
+            "CURVE=bw6_761",
+        ],
+        ":grumpkin_field_id": [
+            "FIELD_ID=GRUMPKIN",
+            "FIELD_ID=GRUMPKIN",
+            "CURVE=grumpkin",
+        ],
+        ":baby_bear_field_id": ["FIELD_ID=BABY_BEAR"],
```
[the remark this answers is not in the document]
Comment (chokobole) [reply third_party/icicle/icicle.BUILD:61]: Yes `BabyBear` doesn't have ones.

**Review by ashjeong — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by GiDeokKim — APPROVED** [review]
LGTM

**Pull request merged**